File: Day 21/Solution.py
# 234. Palindrome Linked List
# Given the head of a singly linked list, return true if it is a palindrome or false otherwise.
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        fast = slow = head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        prev = None
        while slow:
            np = slow.next
            slow.next = prev
            prev = slow
            slow = np

        l = head
        r = prev
        while r:
            if l.val != r.val:
                return False
            l = l.next
            r = r.next
        return True

if __name__ == "__main__":
    # Palindrome: 1 -> 2 -> 2 -> 1
    head = ListNode(1)
    head.next = ListNode(2)
    head.next.next = ListNode(2)
    head.next.next.next = ListNode(1)
    sol = Solution()
    print(sol.isPalindrome(head)) # Output: True

    # Non-palindrome: 1 -> 2 -> 3
    head2 = ListNode(1)
    head2.next = ListNode(2)
    head2.next.next = ListNode(3)
    print(sol.isPalindrome(head2)) # Output: False

File: Day 21/Solution.java
// 234. Palindrome Linked List
// Given the head of a singly linked list, return true if it is a palindrome or false otherwise.
class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

public class Solution {
    public boolean isPalindrome(ListNode head) {
        ListNode fast = head, slow = head;
        while (fast != null && fast.next != null){
            slow = slow.next;
            fast = fast.next.next;
        }

        ListNode prev = null;
        while (slow != null){
            ListNode np = slow.next;
            slow.next = prev;
            prev = slow;
            slow = np;
        }

        ListNode l = head, r = prev;
        while(r != null){
            if (l.val != r.val) {
                return false;
            }
            l = l.next;
            r = r.next;
        }
        return true;
    }

    // Test harness for VS Code
    public static void main(String[] args){
        // Palindrome: 1 -> 2 -> 2 -> 1
        ListNode head = new ListNode(1);
        head.next = new ListNode(2);
        head.next.next = new ListNode(2);
        head.next.next.next = new ListNode(1);

        Solution sol = new Solution();
        System.out.println(sol.isPalindrome(head)); // Output: true

        // Non-palindrome: 1 -> 2 -> 3
        ListNode head2 = new ListNode(1);
        head2.next = new ListNode(2);
        head2.next.next = new ListNode(3);
        System.out.println(sol.isPalindrome(head2)); // Output: false
    }
}
